vendor/plist-to-html/plist-to-html/PlistToHtml

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env python
+
+import argparse
+import json
+import os
+import plistlib
+from xml.parsers.expat import ExpatError
+
+
+def get_report_data_from_plist(plist):
+    files = plist['files']
+
+    reports = []
+    file_sources = {}
+    for diag in plist['diagnostics']:
+        bug_path_items = [item for item in diag['path']]
+        events = [i for i in bug_path_items if i.get('kind') == 'event']
+
+        report = []
+        for index, event in enumerate(events):
+            file_id = event['location']['file']
+            if file_id not in file_sources:
+                file_path = files[file_id]
+                source_data = open(file_path, 'r')
+                file_sources[file_id] = {'id': file_id,
+                                         'path': file_path,
+                                         'content': source_data.read()}
+
+            report.append({'line': event.location['line'],
+                           'col':  event.location['col'],
+                           'file': event.location['file'],
+                           'msg':  event.message,
+                           'step': index + 1})
+        reports.append(report)
+
+    return {'files': file_sources,
+            'reports': reports}
+
+
+def generate_html_from_plist(file_path, plist, layout_dir, output_path):
+    data = get_report_data_from_plist(plist)
+
+    if data is None or not len(data['reports']):
+        print('No report data in ' + file_path + ' file.')
+        return;
+
+    html_output_path = os.path.join(output_path,
+                                    os.path.basename(file_path) + '.html')
+    html_output = open(html_output_path, 'w+')
+
+    header_path = os.path.join(layout_dir, 'header.html')
+    footer_path = os.path.join(layout_dir, 'footer.html')
+
+    if not os.path.isfile(header_path) or not os.path.isfile(footer_path):
+        print('Failed to open ' + header_path + ' and ' + footer_path)
+        exit(1)
+
+    html_header = open(header_path, 'r')
+    html_footer = open(footer_path, 'r')
+
+    html_output.write(html_header.read())
+
+    # Css files
+    codemirror_css = open(os.path.join(layout_dir, 'codemirror.min.css'), 'r')
```

Comment:
None of the files were closed after the read().

Reply:
Couldn't we read the js and css files only one? Not for each html report generation.